staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -986,6 +1026,52 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	return allErrs
 }
 
+func extractMaxElements(schema *apiextensions.JSONSchemaProps) int64 {
+	switch schema.Type {
+	case "object":
+		if schema.AdditionalProperties != nil {
+			return *schema.MaxProperties
+		}
+		// return 1, and not unboundedSchemaLength, because even if we're
+		// not looking at a map here, we can assume the object still preserves
+		// the cardinality of its parents
+		return 1
+	case "array":
+		if schema.MaxItems != nil {
+			return *schema.MaxItems
+		}
+		return unboundedSchemaLength
+	case "string":
+		if schema.MaxLength != nil {
+			return *schema.MaxLength
+		}
+		return unboundedSchemaLength
+	default:
+		// TODO(DangerOnTheRanger): should we give XIntOrString its own case?
+		return unboundedSchemaLength
+	}
+}
+
+func getCRDCost(baseCost uint64, schemaNode *schemaTree) uint64 {
```

Comment:
Maybe use a name different than "CRD cost"?  This computes the cost of a single CEL expression, not the cost of all the CEL expressions in a CRD.